### ISSUE TITLE
Avoid globals in the typescript type definitions

### ISSUE THREAD

**Issue by ggilles** [wix/Detox#4223]
### Describe your idea

Currently, the Detox typescript definitions start with something like this:

```typescript
declare global {
    const detox: Detox.DetoxExportWrapper;
    const device: Detox.DetoxExportWrapper['device'];
    const element: Detox.DetoxExportWrapper['element'];
    const waitFor: Detox.DetoxExportWrapper['waitFor'];
    // ...
```

This makes those declarations available in the global namespace.

However, it still required to:

```typescript
import { by, element, expect, waitFor, device } from 'detox';
```

Because just declaring those functions as global doesn't make them available at runtime in the context of typescript compilation.

For example: 
```typescript
describe('Example', () => {
    it('runtime fail !', async () => {
        await device.launchApp();
    });
});
```
will fail at runtime like this:
```
  Example
    ✕ runtime fail ! (7929 ms)

  ● Example › runtime fail !

    ReferenceError: device is not defined

      1 | describe('Example', () => {
      2 |     it('runtime fail !', async () => {
    > 3 |         await device.launchApp();
        |         ^
      4 |     });
      5 | });
      6 |
```

The reason it fails is because we are missing the proper `import { device } from 'detox';` of course.

The problem here is that the **presence of the global declaration of `device`** prevents proper imports resolution from happening correctly: VS code will not suggest any missing import, and `tsc` will not complain about it either. This is an annoying usability issue.

**Comment by noomorph** [comment]
By the way, it looks like you have some local configuration issue there. For the most part the typings should work, the only known issue is `expect`. Do you have skipLibCheck: true? Try to run your TypeScript with --listFiles when without explicit detox import, and see if it picks our index.d.ts at all. It should, actually.

**Comment by ggilles** [comment]
@noomorph Well, we actually tried doing the setup several times 🤔 

Again, the typing **does work**, but the **automatic import** does not.

For example, picking https://github.com/ixartz/React-Native-Boilerplate, here is a video of what happens : in the video I am first demonstrating what should happen when import resolution works fine by editing the `Welcome.test.tsx` file, and then I edit a detox e2e test to show the problem: in this case, the `element` import is not properly added back.

https://github.com/wix/Detox/assets/1478233/297625f3-0748-4a70-943b-ff531bca064a

I had a look and tried several repositories, including https://github.com/software-mansion/react-native-screens/tree/main/Example which is fairly recent and they all show the same problematic behavior.

If you do have a link to a working example, can you please share it ?

**Comment by noomorph** [comment]
Well, the behavior shown on the video is rather expected since those imports are equally visible as global variables.

Indeed, if you were using a special Detox setup with initGlobals: false, then this could be an issue.

As for the example, I think our examples/demo-react-native project should have TS files. Sorry, I am not near the computer.

If I have time tomorrow, I might take a look at the linked project and see what you mean.

Nevertheless, even if we split index.d.ts into 2: index and globals - it still requires a major version bump. I think this makes sense overall, but I'd need to play with various TS configurations to see what works the best. Or you could be the one to do that. :wink:

**Comment by ggilles** [comment]
> Well, the behavior shown on the video is rather expected since those imports are equally visible as global variables.

That's the point of the issue: those imports are visible as global variables so automatic import does not work.

> As for the example, I think our examples/demo-react-native project should have TS files. Sorry, I am not near the computer.

I tried that example, the problem exists there too.

> Nevertheless, even if we split index.d.ts into 2: index and globals - it still requires a major version bump. I think this makes sense overall, but I'd need to play with various TS configurations to see what works the best. Or you could be the one to do that.

I really don't understand why this would be a breaking change:  if the import aren't done explicitly then it does not work and fails at runtime. What are we going to break if it doesn't work anyway ?

**Comment by ggilles** [comment]
@noomorph When I said "I tried that example, the problem exists there too..." I was talking about the **automatic import resolution does not work**, sorry if this wasn't clear enough.

Now, for that part:

> if the import aren't done explicitly then it does not work and fails at runtime

This is what happens in our project and probably where something is off. You mentioned `initGlobals: false` and I couldn't find it, but I realize it is actually named `exposeGlobals: false`. In our project, we indeed use `exposeGlobals: false` to prevent the hijacking of the jest definition of `expect` by `Detox`. And then we have come full circle and I finally understand your initial remark about it.

So, I have no idea how to solve this in a proper way. The root cause of the issue is in fact that when using `exposeGlobals: false` the typescript global definition should not be there.

**Comment by noomorph** [comment]
I agree that this is a problem, but even then, the solution will be to provide non-intrusive typings along with the optional intrusive ones.

This situation somewhat reminds me of Jest.

They have `@jest/types`, but there is also `@types/jest` living its parallel life in the universe. The difference between those is that the former is self-contained, while `@types/jest` augments the global scope.

I'm okay with addressing your concern in the next major release but not in a minor one. If I release a version with your patch, it will break tests where they have:

```ts
import { expect } from 'detox';

device // @ts-expect-error 💥 
```

I know what I'm saying - there are about 8-9 repositories at Wix alone with omitted imports I know personally, and there are even more unknown to me. A necessity of any kind of migration <==> major release.

Your pull request at this stage lacks a migration plan for such projects. By migration, I mean dead simple - a) install this or that legacy adapter b) edit something in your config - et voila, it works again.

I don't see the solution in your version for that anyway, so more work has to be done, in my opinion.

The bare minimum is to create that extra global typings file, and update our TypeScript guide how to make it work.

**Comment by noomorph** [comment]
Related #2610.

**Comment by ggilles** [comment]
@noomorph Thanks for that rather complete explanation. 🙏

**Comment by noomorph** [comment]
Nonetheless, I will attempt soon to extract globals.d.ts from index.d.ts and begin this journey, at the very least.

**Comment by noomorph** [comment]
@ggilles, I created a PR, #4243, to help you a bit with your issue:

```ts
/// <reference path="detox.d.ts" />
/// <reference path="globals.d.ts" />

declare const detox: Detox.DetoxExportWrapper;
export = detox;
```

↑ This is what `node_modules/detox/index.d.ts` will look like.

You'll be able to delete `/// <reference path="globals.d.ts" />` manually, which should alleviate your autocompletion issue.

This is obviously a workaround, but I hope it makes your test writing easier.

I think that we'll be able to ship Detox 21 (but not earlier) without globals included by default so that the users would need to re-enable them again manually, for example, like this:

```diff
   "compilerOptions": {
    ...
-     "types": ["node", "jest", "detox"]
+     "types": ["node", "jest", "detox/globals"]
   },
```

**Comment by ggilles** [comment]
@noomorph Thank you so much, I didn't expect to see this problem to be addressed so fast.

**Comment by noomorph** [comment]
@ggilles the version with the workaround above will be released within the next 30 minutes. It should be 20.13.3

**Comment by noomorph** [comment]
> Thank you so much, I didn't expect to see this problem to be addressed so fast.

Well, I'm happy to prioritize issues of proactive people who contribute to Detox as much as I can, even in the context of busy schedule and 75% of team in low availability.

I hope that Asaf manages soon enough to address that issue with swizzling that "cancelSomething" in Objective-C. I think you did a brilliant investigation back at the time.

**Comment by noomorph** [comment]
20.13.3 is out, you can try now that suggestion, @ggilles. Let me know if removing that line manually fixes your autocompletion issues.

**Comment by noomorph** [comment]
This item has been fixed in https://github.com/wix/Detox/pull/3975/commits/8e8e363fda597e3e22e1a99fdc1ce9179626aec6 and it will be released with Detox 21.
